File: docs/research/corpus-examples/file-handling/python/vulnerable-4-altered.py
# vulnerable-4-altered.py
# MANUFACTURED per docs/VULN_CORPUS_PAIR_MANUFACTURING_PLAN.md's "Per-pair
# mechanics". A distinct upload idiom from this file's Flask/werkzeug
# pair (vulnerable-1.py/idiomatic-1.py): a FastAPI route. Unrestricted
# upload: the client-declared content_type is never checked, and the
# file is written to disk using the client-supplied original filename
# verbatim.
from fastapi import APIRouter, UploadFile

router = APIRouter()
UPLOAD_DIR = "/srv/app/public/uploads"


@router.post("/upload/profile-doc")
async def upload_profile_doc(file: UploadFile):
    contents = await file.read()
    dest = f"{UPLOAD_DIR}/{file.filename}"
    with open(dest, "wb") as f:
        f.write(contents)
    return {"status": "uploaded", "path": dest}

File: docs/research/corpus-examples/file-handling/php/vulnerable-4-altered.php
<?php
// vulnerable-4-altered.php
// MANUFACTURED per docs/VULN_CORPUS_PAIR_MANUFACTURING_PLAN.md's "Per-pair
// mechanics". A distinct upload-validation idiom from this file's
// extension+finfo-allowlist pair (vulnerable-1.php/idiomatic-1.php):
// avatar upload that trusts the client-declared Content-Type of the
// multipart part alone (never sniffing or re-encoding the actual image
// bytes) before saving the file with its original extension.
$type = $_FILES['avatar']['type'] ?? '';

if (strpos($type, 'image/') === 0) {
    $ext = substr($type, strlen('image/'));
    $dest = __DIR__ . '/avatars/' . uniqid('avatar_', true) . '.' . $ext;
    move_uploaded_file($_FILES['avatar']['tmp_name'], $dest);
    echo json_encode(['status' => 'uploaded', 'path' => $dest]);
} else {
    http_response_code(415);
    echo json_encode(['error' => 'unsupported type']);
}
